libraries/AP_HAL_ChibiOS/AnalogIn.cpp

```diff
@@ -289,6 +293,119 @@ void AnalogIn::read_adc(uint32_t *val)
     chSysUnlock();
 }
 
+
+#if HAL_WITH_MCU_MONITORING
+/*
+  on H7 we can support monitoring MCU temperature and voltage using ADC3
+ */
+#define ADC3_GRP1_NUM_CHANNELS 3
+
+// internal ADC channels (from H7 reference manual)
+#define ADC3_VSENSE_CHAN 18
+#define ADC3_VREFINT_CHAN 19
+#define ADC3_VBAT4_CHAN 17
+
+// samples filled in by ADC DMA engine
+adcsample_t *AnalogIn::samples_adc3;
+uint32_t AnalogIn::sample_adc3_sum[ADC3_GRP1_NUM_CHANNELS];
+// we also keep min and max so we can report the range of voltages
+// seen, to give an idea of supply stability
+uint16_t AnalogIn::sample_adc3_max[ADC3_GRP1_NUM_CHANNELS];
+uint16_t AnalogIn::sample_adc3_min[ADC3_GRP1_NUM_CHANNELS];
+uint32_t AnalogIn::sample_adc3_count;
+
+/*
+  callback from ADC3 driver when sample buffer is filled
+ */
+void AnalogIn::adc3callback(ADCDriver *adcp)
+{
+    const adcsample_t *buffer = samples_adc3;
+
+    stm32_cacheBufferInvalidate(buffer, sizeof(adcsample_t)*ADC_DMA_BUF_DEPTH*ADC3_GRP1_NUM_CHANNELS);
+    for (uint8_t i = 0; i < ADC_DMA_BUF_DEPTH; i++) {
+        for (uint8_t j = 0; j < ADC3_GRP1_NUM_CHANNELS; j++) {
+            const uint16_t v = *buffer++;
+            sample_adc3_sum[j] += v;
+            if (sample_adc3_min[j] == 0 ||
+                sample_adc3_min[j] > v) {
+                sample_adc3_min[j] = v;
+            }
+            if (sample_adc3_max[j] == 0 ||
+                sample_adc3_max[j] < v) {
+                sample_adc3_max[j] = v;
+            }
+        }
+    }
+    sample_adc3_count += ADC_DMA_BUF_DEPTH;
+}
+
+/*
+  setup ADC3 for internal temperature and voltage monitoring
+ */
+void AnalogIn::setup_adc3(void)
+{
+    samples_adc3 = (adcsample_t *)hal.util->malloc_type(sizeof(adcsample_t)*ADC_DMA_BUF_DEPTH*ADC3_GRP1_NUM_CHANNELS, AP_HAL::Util::MEM_DMA_SAFE);
```

Comment:
not checking if allocation was successful?

Reply:
fixed to not init adc3 if alloc fails